pkg/metrics/metrics.go

```diff
@@ -0,0 +1,239 @@
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/golang/glog"
+	mapiv1beta1 "github.com/openshift/cluster-api/pkg/apis/machine/v1beta1"
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	// MachineCountDesc is a metric about machine object count in the cluster
+	MachineCountDesc = prometheus.NewDesc("mapi_machine_items_total", "Count of machine objects currently at the apiserver", nil, nil)
+	// MachineSetCountDesc Count of machineset object count at the apiserver
+	MachineSetCountDesc = prometheus.NewDesc("mapi_machineset_items_total", "Count of machinesets at the apiserver.", nil, nil)
+
+	//MachineInfo Information of the Machines currently managed by the machine controller.
+	MachineInfo = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mapi_machine_info",
+		Help: "Information of the Machines currently managed by the machine controller.",
+	}, []string{"name", "namespace", "createdAt",
```

Comment:
Are we sure we want things like timestamps and instance IDs as labels in these metrics? These things have a tendency in some cases to create a huge number of individual time series in Prometheus, and in the case of a `createdAt` timestamp it not super useful to query for.

Reply:
May be you are right. Another option is to let these metrics in at this point so that we have basic support working and then refine metrics in follow-up PRs. I am fine either way/
/cc @enxebre

Reply:
I'd keep the `provider id`, also if we rename node_ref to node we could in a follow up have an alert for machines with no nodes/invalid nodes the likes of ```(machine_info unless on(node) kube_node_info) > 0```

Reply:
From my Prometheus developer's POV, `createdAt` smells like something that shouldn't be a label. As @bison said, it won't useful because label values are opaque to Prometheus and you can't do any arithmetic on them. What would work is an additional metric reporting this value (eg `mapi_machine_created_at_seconds`).